File: src/main/java/com/proximyst/schizoparadise/effects/Disc11Effect.java
package com.proximyst.schizoparadise.effects;

import co.aikar.taskchain.TaskChainFactory;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.util.Vector;

import java.util.concurrent.ThreadLocalRandom;

public class Disc11Effect extends Effect {
    private final TaskChainFactory chainFactory;

    public Disc11Effect(TaskChainFactory chainFactory) {
        super("Disc 11");
        this.chainFactory = chainFactory;
    }

    @Override
    public void apply(Player player) {
        Location location = player.getLocation().clone().add(
                new Vector(ThreadLocalRandom.current().nextInt(-5, 5), -3, ThreadLocalRandom.current().nextInt(-5, 5))
        );
        chainFactory.newChain()
                .sync(() -> location.getWorld().playEffect(location, org.bukkit.Effect.RECORD_PLAY, Material.RECORD_11))
                .delay(ThreadLocalRandom.current().nextInt(20, 60))
                .sync(() -> location.getWorld().playEffect(location, org.bukkit.Effect.RECORD_PLAY, 0))
                .execute();
    }
}

File: src/main/kotlin/com/proximyst/schizoparadise/effect/Disc11Effect.kt
package com.proximyst.schizoparadise.effect

import com.proximyst.schizoparadise.Main
import com.proximyst.schizoparadise.data.SchizophrenicPlayer
import org.bukkit.Material
import org.bukkit.metadata.MetadataValueAdapter
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.TimeUnit

class Disc11Effect(private val main : Main) : Effect() {
  override val name : String = "Disc 11"

  override fun apply(player : SchizophrenicPlayer) {
    val location = player.location.clone().add(
        ThreadLocalRandom.current().nextDouble(-5.0, 5.0), -3.0, ThreadLocalRandom.current().nextDouble(-5.0, 5.0)
    )
    main.taskChainFactory
        .newChain<Unit>()
        .sync {
          location.world.playEffect(location, org.bukkit.Effect.RECORD_PLAY, Material.RECORD_11)
          player.addPotionEffect(PotionEffect(PotionEffectType.BLINDNESS, 3000, 1))
          player.setMetadata("Blindness", object : MetadataValueAdapter(main) {
            override fun invalidate() {
            }

            override fun value() : Any {
              return true
            }
          })
        }
        .delay(ThreadLocalRandom.current().nextInt(5, 20), TimeUnit.SECONDS)
        .sync {
          location.world.playEffect(location, org.bukkit.Effect.RECORD_PLAY, 0)
        }
        .delay(10)
        .sync {
          if (player.isOnline) {
            player.removePotionEffect(PotionEffectType.BLINDNESS)
            player.removeMetadata("Blindness", main)
          }
        }
        .execute()
  }
}